### PR TITLE
Begin making examples Python 3 compatible

### PR DESCRIPTION
**Pull request by elitest**
This PR makes a few of the examples work under Python 3 in addition to Python 2.  I ran some tests under 2.7.8 and 3.4.2.

I also found a couple of md5s being used during signing in certgen.py, I'm proposing that they be bumped up to SHA-256.

### CODE REVIEW

File: examples/certgen.py [hunk 1]
```diff
@@ -45,14 +45,14 @@ def createCertRequest(pkey, digest="sha256", **name):
...
-    for (key,value) in name.items():
+    for (key,value) in list(name.items()):
         setattr(subj, key, value)
 
     req.set_pubkey(pkey)
     req.sign(pkey, digest)
     return req
 
-def createCertificate(req, (issuerCert, issuerKey), serial, (notBefore, notAfter), digest="md5"):
+def createCertificate(req, issuerCertKey, serial, validityPeriod, digest="sha256"):
```
Comment (elitest) [inline examples/certgen.py]: According to [PEP 3113](https://www.python.org/dev/peps/pep-3113/) tuple unpacking is deprecated.

File: examples/certgen.py [hunk 2]
```diff
@@ -45,14 +45,14 @@ def createCertRequest(pkey, digest="sha256", **name):
     req = crypto.X509Req()
     subj = req.get_subject()
 
-    for (key,value) in name.items():
+    for (key,value) in list(name.items()):
```
Comment (hynek) [inline examples/certgen.py]: Why do you convert them into a list? Please don’t. :)

Also let’s just write `key, value` instead of `(key,value)`.

File: examples/certgen.py [hunk 3]
```diff
@@ -64,9 +64,11 @@ def createCertificate(req, (issuerCert, issuerKey), serial, (notBefore, notAfter
                             starts being valid
                notAfter   - Timestamp (relative to now) when the certificate
                             stops being valid
-               digest     - Digest method to use for signing, default is md5
+               digest     - Digest method to use for signing, default is sha256
     Returns:   The signed certificate in an X509 object
     """
+    (issuerCert, issuerKey) = issuerCertKey
+    (notBefore, notAfter) = validityPeriod
```
Comment (hynek) [inline examples/certgen.py]: again, no need for parens

File: examples/mk_simple_certs.py [hunk 4]
```diff
@@ -7,15 +7,18 @@
 cakey = createKeyPair(TYPE_RSA, 2048)
 careq = createCertRequest(cakey, CN='Certificate Authority')
 cacert = createCertificate(careq, (careq, cakey), 0, (0, 60*60*24*365*5)) # five years
+
 print('Creating Certificate Authority private key in "simple/CA.pkey"')
-open('simple/CA.pkey', 'w').write(crypto.dump_privatekey(crypto.FILETYPE_PEM, cakey))
+open('simple/CA.pkey', 'w').write(crypto.dump_privatekey(crypto.FILETYPE_PEM, cakey).decode('utf-8'))
```
Comment (hynek) [inline examples/mk_simple_certs.py]: this should be converted  into a context manager

File: examples/mk_simple_certs.py [hunk 5]
```diff
@@ -7,15 +7,18 @@
...
 print('Creating Certificate Authority certificate in "simple/CA.cert"')
-open('simple/CA.cert', 'w').write(crypto.dump_certificate(crypto.FILETYPE_PEM, cacert))
+open('simple/CA.cert', 'w').write(crypto.dump_certificate(crypto.FILETYPE_PEM, cacert).decode('utf-8'))
```
Comment (hynek) [inline examples/mk_simple_certs.py]: this should be converted  into a context manager

File: examples/mk_simple_certs.py [hunk 6]
```diff
@@ -7,15 +7,18 @@
...
+
 for (fname, cname) in [('client', 'Simple Client'), ('server', 'Simple Server')]:
     pkey = createKeyPair(TYPE_RSA, 2048)
     req = createCertRequest(pkey, CN=cname)
     cert = createCertificate(req, (cacert, cakey), 1, (0, 60*60*24*365*5)) # five years
+
     print('Creating Certificate %s private key in "simple/%s.pkey"' % (fname, fname))
-    open('simple/%s.pkey' % (fname,), 'w').write(crypto.dump_privatekey(crypto.FILETYPE_PEM, pkey))
+    open('simple/%s.pkey' % (fname,), 'w').write(crypto.dump_privatekey(crypto.FILETYPE_PEM, pkey).decode('utf-8'))
```
Comment (hynek) [inline examples/mk_simple_certs.py]: this should be converted  into a context manager

File: examples/mk_simple_certs.py [hunk 7]
```diff
@@ -7,15 +7,18 @@
...
     print('Creating Certificate %s certificate in "simple/%s.cert"' % (fname, fname))
-    open('simple/%s.cert' % (fname,), 'w').write(crypto.dump_certificate(crypto.FILETYPE_PEM, cert))
+    open('simple/%s.cert' % (fname,), 'w').write(crypto.dump_certificate(crypto.FILETYPE_PEM, cert).decode('utf-8'))
```
Comment (hynek) [inline examples/mk_simple_certs.py]: this should be converted  into a context manager.

File: examples/simple/client.py [hunk 8]
```diff
@@ -13,11 +13,11 @@
 
 def verify_cb(conn, cert, errnum, depth, ok):
     # This obviously has to be updated
-    print 'Got certificate: %s' % cert.get_subject()
+    print('Got certificate: %s' % cert.get_subject())
```
Comment (hynek) [inline examples/simple/client.py]: please make the % arguments a tuple since you’re touching the line

File: examples/simple/server.py [hunk 9]
```diff
@@ -14,11 +14,11 @@
 
 def verify_cb(conn, cert, errnum, depth, ok):
     # This obviously has to be updated
-    print 'Got certificate: %s' % cert.get_subject()
+    print('Got certificate: %s' % cert.get_subject())
```
Comment (hynek) [inline examples/simple/server.py]: same here, tuple please

File: examples/simple/server.py [hunk 10]
```diff
@@ -57,14 +57,14 @@ def dropClient(cli, errors=None):
 
 while 1:
     try:
-        r,w,_ = select.select([server]+clients.keys(), writers.keys(), [])
+        r,w,_ = select.select([server]+list(clients.keys()), list(writers.keys()), [])
```
Comment (hynek) [inline examples/simple/server.py]: please add missing spaces after `,` and around `+`

File: examples/simple/server.py [hunk 11]
```diff
@@ -88,13 +88,13 @@ def dropClient(cli, errors=None):
...
-        except SSL.Error, errors:
+        except SSL.Error as errors:
             dropClient(cli, errors)
         else:
             writers[cli] = writers[cli][ret:]
             if writers[cli] == '':
                 del writers[cli]
 
-for cli in clients.keys():
+for cli in list(clients.keys()):
```
Comment (hynek) [inline examples/simple/server.py]: don’t convert it to a list please.

File: examples/mk_simple_certs.py (hunk 4 above)
Comment (elitest) [inline examples/mk_simple_certs.py]: Thanks for the feedback @hynek.  I thought contexts were for creating connections with certs that had already been created.  Or am I missing something?

Comment (alex) [inline examples/mk_simple_certs.py]: He meant a context manager:

```py
with open(...) as f:
    f.write(...)
```

On Wed, Apr 29, 2015 at 10:25 PM, elitest <notifications@github.com> wrote:

> In examples/mk_simple_certs.py
> <https://github.com/pyca/pyopenssl/pull/245#discussion_r29399782>:
>
> >  print('Creating Certificate Authority private key in "simple/CA.pkey"')
> > -open('simple/CA.pkey', 'w').write(crypto.dump_privatekey(crypto.FILETYPE_PEM, cakey))
> > +open('simple/CA.pkey', 'w').write(crypto.dump_privatekey(crypto.FILETYPE_PEM, cakey).decode('utf-8'))
>
> Thanks for the feedback @hynek <https://github.com/hynek>. I thought
> contexts were for creating connections with certs that had already been
> created. Or am I missing something?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/pyopenssl/pull/245/files#r29399782>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

File: examples/simple/server.py [hunk 12]
```diff
@@ -8,17 +8,19 @@
...
+from OpenSSL import SSL, crypto
 import sys, os, select, socket
 
 
 def verify_cb(conn, cert, errnum, depth, ok):
     # This obviously has to be updated
-    print 'Got certificate: %s' % cert.get_subject()
+    certsubject = crypto.X509Name(cert.get_subject())
+    commonname = certsubject.commonName
+    print(('Got certificate: ' + commonname))
```
Comment (alex) [inline examples/simple/server.py]: Why are these changes needed, particularly the `crypto.X509Name(...)` bit?

Comment (elitest) [inline examples/simple/server.py]: I should have made a comment to clarify :).

This is what you get on master:
`Got certificate: <X509Name object '/CN=Certificate Authority'>`
`Got certificate: <X509Name object '/CN=Simple Server'>`

This is what you get with that:
`Got certificate: Certificate Authority`
`Got certificate: Simple Server`

I think that is the direction the orig author was going in when he said: "# This obviously has to be updated".  I can remove the comment if you agree.  I suppose there are less spoofable things that we could use to identify the certs in addition to the CN, like the fingerprint.

File: examples/certgen.py (hunk 2 above)
Comment (elitest) [inline examples/certgen.py]: Fixed in 473fe6a

File: examples/certgen.py (hunk 3 above)
Comment (elitest) [inline examples/certgen.py]: Fixed in 473fe6a

File: examples/mk_simple_certs.py (hunk 4 above)
Comment (elitest) [inline examples/mk_simple_certs.py]: Fixed in aab9ddd

File: examples/mk_simple_certs.py (hunk 5 above)
Comment (elitest) [inline examples/mk_simple_certs.py]: Fixed in aab9ddd

File: examples/mk_simple_certs.py (hunk 6 above)
Comment (elitest) [inline examples/mk_simple_certs.py]: Fixed in aab9ddd

File: examples/mk_simple_certs.py (hunk 7 above)
Comment (elitest) [inline examples/mk_simple_certs.py]: Fixed in aab9ddd

File: examples/simple/client.py (hunk 8 above)
Comment (elitest) [inline examples/simple/client.py]: Fixed in b2ff5be

File: examples/simple/server.py (hunk 9 above)
Comment (elitest) [inline examples/simple/server.py]: Fixed in b2ff5be

File: examples/simple/server.py (hunk 10 above)
Comment (elitest) [inline examples/simple/server.py]: lists removed(4852ef4) and then re-added(b2ff5be) as will not work under python3.  Spacing fixed in those commits.

File: examples/simple/server.py (hunk 11 above)
Comment (elitest) [inline examples/simple/server.py]: Fixed in 473fe6a

File: examples/certgen.py (hunk 1 above)
Comment (hynek) [inline examples/certgen.py]: it’s actually even *removed* in Python 3 (which kind of sucks when doing Twisted)

File: examples/simple/server.py (hunk 12 above)
Comment (hynek) [inline examples/simple/server.py]: yeah please get rid of the comment.  there’s nothing obvious about it.

**Comment by elitest** [comment]
Thanks to @hynek's work in #253 this now passes.  Cheers.

**Comment by hynek** [comment]
So, stupid question: is there currently an easy way to actually test the examples?

**Comment by elitest** [comment]
the 'easy' way I used to test them from the examples directory was:
run 'python mk_simple_certs.py'
run 'python simple/server.py 8443'
run 'python simple/client.py localhost 8443'

then type some text in, and server.py should echo it back to you, as well as tell you what certificate it received.  It actually just gives the CN as opposed to something that couldn't be spoofed like a fingerprint.  Then I ran the same against python3.

**Comment by hynek** [comment]
I see, thanks!

Please add `examples/simple/*.cert` and `examples/simple/*.pkey` to .gitignore then.

**Comment by hynek** [comment]
Thanks!
